apps/aecore/lib/aecore/peers/sync.ex

```diff
@@ -1,44 +1,121 @@
+                                       chain_sync_running: false},
+                         name: __MODULE__)
   end
 
   def init(state) do
     {:ok, state}
   end
 
-  def get_state() do
-    GenServer.call(__MODULE__, :get_state)
+  @spec get_peer_blocks() :: map()
+  def get_peer_blocks() do
+    GenServer.call(__MODULE__, :get_peer_blocks)
+  end
+
+  @spec add_running_task(String.t()) :: :ok
+  def add_running_task(peer_uri) do
+    GenServer.call(__MODULE__, {:add_running_task, peer_uri})
+  end
+
+  @spec remove_running_task(String.t()) :: :ok
+  def remove_running_task(peer_uri) do
+    GenServer.call(__MODULE__, {:remove_running_task, peer_uri})
+  end
+
+  @spec get_running_tasks() :: map()
+  def get_running_tasks() do
+    GenServer.call(__MODULE__, :get_running_tasks)
+  end
+
+  @spec get_chain_sync_status() :: boolean()
+  def get_chain_sync_status() do
+    GenServer.call(__MODULE__, :get_chain_sync_status)
   end
 
-  @spec add_block_to_state(binary, term()) :: :ok
+  @spec set_chain_sync_status(boolean()) :: :ok
+  def set_chain_sync_status(status) do
+    GenServer.call(__MODULE__, {:set_chain_sync_status, status})
+  end
+
+  @spec add_block_to_state(binary(), term()) :: :ok
   def add_block_to_state(block_hash, block) do
     GenServer.call(__MODULE__, {:add_block_to_state, block_hash, block})
   end
 
+  @spec remove_block_from_state(binary()) :: :ok
+  def remove_block_from_state(block_hash) do
+    GenServer.call(__MODULE__, {:remove_block_from_state, block_hash})
+  end
+
   @spec ask_peers_for_unknown_blocks(map()) :: :ok
   def ask_peers_for_unknown_blocks(peers) do
-    GenServer.call(__MODULE__, {:ask_peers_for_unknown_blocks, peers})
+    Enum.each(peers, fn ({_, %{uri: uri, latest_block: top_block_hash}}) ->
+      {:ok, top_hash_decoded} = Base.decode16(top_block_hash)
+      if !Map.has_key?(get_running_tasks(), uri) do
+        PeerBlocksTask.start_link([uri, top_hash_decoded])
+      end
+    end)
   end
 
-  @spec add_valid_peer_blocks_to_chain() :: :ok
-  def add_valid_peer_blocks_to_chain() do
-    GenServer.call(__MODULE__, :add_valid_peer_blocks_to_chain)
+  @spec add_peer_blocks_to_sync_state(String.t(), binary()) :: :ok
+  def add_peer_blocks_to_sync_state(peer_uri, from_block_hash) do
+    if !Chain.has_block?(from_block_hash) do
+      add_running_task(peer_uri)
+      case HttpClient.get_raw_blocks({peer_uri, from_block_hash, Chain.top_block_hash()}) do
+        {:ok, blocks} ->
+          if !Enum.empty?(blocks) do
+            Enum.each(blocks, fn(block) ->
+              try do
+                BlockValidation.single_validate_block!(block)
+                peer_block_hash = BlockValidation.block_header_hash(block.header)
+                if !Chain.has_block?(peer_block_hash) do
+                  add_block_to_state(peer_block_hash, block)
+                end
+              catch
+                {:error, message} ->
+                  Logger.error(fn -> message end)
```

Comment:
When a sync with some peer fails and `Logger.error` is called `remove_running_task(peer_uri)` won't get called. This will cause the peer to be excluded from every future block sync, because `Map.has_key?(get_running_tasks(), uri)` will always be true.